skbio/io/format/stockholm.py

```diff
@@ -0,0 +1,301 @@
+...         'M24803.1/1-23                  UAAGUUCUCGAUCUCUAAAAUCG',
+...         '#=GC SS_cons                   .AAA....<<<<aaa....>>>>',
+...         '//'
+...     ])
+>>> fh = StringIO(fs)
+>>> msa = skbio.io.read(fh, into=TabularMSA, constructor=RNA)
+>>> msa # doctest: +NORMALIZE_WHITESPACE
+TabularMSA[RNA]
+-------------------------------------------------
+Metadata:
+'RA': '   Deiman BA, Kortlever RM, Pleij CW;'
+'RL': '   J Virol 1997;71:5990-5996.'
+Positional metadata:
+'SS_cons': <dtype: object>
+Stats:
+sequence count: 4
+position count: 23
+-------------------------------------------------
+UGAGUUCUCGAUCUCUAAAAUCG
+UGAGUUCUCUAUCUCUAAAAUCG
+UAAGUUCUCGAUCUUUAAAAUCG
+UAAGUUCUCGAUCUCUAAAAUCG
+
+References
+==========
+.. [1] https://en.wikipedia.org/wiki/Stockholm_format
+.. [2] http://sonnhammer.sbc.su.se/Stockholm.html
+"""
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+from collections import OrderedDict
+
+from skbio.alignment import TabularMSA
+from skbio.sequence import Protein
+from skbio.io import create_format, StockholmFormatError
+
+stockholm = create_format('stockholm')
+
+
+@stockholm.sniffer()
+def _stockholm_sniffer(fh):
+    # Smells a Stockholm file if the first line contains 'Stockholm' and the
+    # version number.
+    line = next(fh)
+    if line[:15] == "# STOCKHOLM 1.0":
+        return True, {}
+
+    return False, {}
+
+
+@stockholm.reader(TabularMSA)
+def _stockholm_to_tabular_msa(fh, constructor=Protein):
+    # Uses OrderedDict() to make sure dna_data isn't arranged randomly
+    dna_data = OrderedDict()
+    metadata = {}
+    positional_metadata = {}
+    seqs = []
+
+    # Retrieves data from file
+    for line in fh:
+        if not (line.startswith("#") or line.startswith("//") or
+                line.isspace() == True):
+            dna_data = _parse_stockholm_line_data(line, dna_data)
+    fh.seek(0)
+
+    # Retrieves metadata from file
+    for line in fh:
+        if line.startswith("#=GF"):
+            metadata = _parse_stockholm_line_gf(line, metadata)
+        if line.startswith("#=GS"):
+            dna_data = _parse_stockholm_line_gs(line, dna_data)
+        if line.startswith("#=GR"):
+            dna_data = _parse_stockholm_line_gr(line, dna_data)
+        if line.startswith('#=GC'):
+            positional_metadata = _parse_stockholm_line_gc(line,
+                                                           positional_metadata)
+
+    for key in dna_data.keys():
+        # Sets blank dictionaries and lists to None instead
+        if not dna_data[key][1]:
```

Comment:
I recommend using temp variables to clarify what these keys and indices represent. For example:

```python
metadata = dna_data[key][1]
positional_metadata = dna_data[key][2]

if not metadata:
    ...
```

This is more efficient since it doesn't have to perform the index lookup multiple times, but more importantly makes the code easier to read.